docs/guides/implementation-guides/how-to-implement-dependency-injection.md

```diff
@@ -7,9 +7,120 @@ import MvvmArchitectureDiagram from '/img/gitbook-import/assets/image (3) (1) (2
 
 # How To Implement Dependency Injection
 
-Content in preparation.
+:::info
+For views you can make a simple [ViewLocator](../../concepts/view-locator) which will automatically resolve the view from the Locator.
+:::
 
 This guide will show you how to use Dependency Injection (DI) with _Avalonia UI_ and the MVVM pattern. 
 
 <img src={MvvmArchitectureDiagram} alt=""/>
 
+Let's assume that your MainViewModel as a dependency on MyService. The contructor for ViewModel should looks like this:
+```csharp
+private readonly MyService _myService;
+
+public MapsViewModel(MyService myService)
+{
+    _myService = myService;
+}
+```
+These are the steps you need to do in order to resolve that dependencu using DI.
+
+## Step 1: Install the NuGet package Microsoft.Extensions.DependencyInjection
+```shell
+dotnet add package Microsoft.Extensions.DependencyInjection
+```
+
+## Step 2: ServiceCollectionExtensions 
+The following code is creating an extension for IServiceCollection. This is where you will register all your common dependencies. It's recomended to create that class in the shared project used by all avalonia target platform.
+
+```csharp
+public static class ServiceCollectionExtensions {
+    public static void AddCommonServices(this IServiceCollection collection) {
+        collection.AddTransient<MainViewModel>();
+        collection.AddTransient<MyService>();
+    }
+}
+```
+
+## Step 3: ServiceManager
+The following code create a ServiceManager class that is used to retrieve an instance of a given object.
+```csharp
+public static class ServiceManager {
+    public static IServiceProvider Locator { get; private set; }
+
+    public static void InitialiseLocatorFromCollection(IServiceCollection collection) {
+        Locator = collection.BuildServiceProvider();
```

Comment:
This code was fine for me to just quickly show off how to do this, but since people are probably going to copy and paste this a lot it's worth doing properly. As such I would suggest:

1. Add comments to the `Locator` property and the `InitialiseLocatorFromCollection`method
2. Make the `InitialiseLocatorFromCollection` method check if the locator has already been initialised
3. Add some functionality to dispose of the locator when the platform shutsdown as people may have dependencies that need to be properly cleaned up

Reply:
I've just done 2. but will need your help for 1. and 3.

Reply:
1.  ```csharp
    /// <summary>
    /// Handles global dependency injection related functionality
    /// </summary>
    public static class ServiceManager {
    /// <summary>
    /// A singleton instance of the service provider
    /// </summary>
    /// <remarks>
    /// This can be used to directly resolve services from the container. However doing so should be avoided whenever
    /// possible to avoid creating a dependency on the locator instance itself. In the vast majority of cases, the service
    /// container will be able automatically resolve services for you 
    /// </remarks>
    public static IServiceProvider? Locator { get; private set; }

    /// <summary>
    /// Initialise <see cref="Locator"/> by building the <paramref name="collection"/> into a <see cref="IServiceProvider"/>
    /// </summary>
    /// <param name="collection">The collection to create a service provider from</param>
    /// <remarks>This method may only ever be called ONCE</remarks>
    public static void InitialiseLocatorFromCollection(IServiceCollection collection) {
        //Check if the locator has been initialised previously
        if (Locator is null) {
            //Build the locator from the given collection
            Locator = collection.BuildServiceProvider();
            return;
        }

        //The locator has already been set, throw an error since this may lead to an inconsistent application state if
        //dependencies have already been resolved from the old locator
        throw new ApplicationException("The locator cannot be set more than once without explicitly disposing it!");
    }
    }

    /// <summary>
    /// Disposes the <see cref="Locator"/> instance and sets it back to a state where it can be reinitialised
    /// </summary>
    /// <remarks>This method will typically only ever be called on an application exit</remarks>
    public static void DisposeLocator() {
        if (Locator is not null && Locator is IDisposable disposable) {
            disposable.Dispose();
            Locator = null;
        }
    }
    ```
2. Put this somewhere that will be executed during boot where you have access to the Application lifetime (EG app class). It wont always work since you can't run code if something kills your app, but it should cover most graceful exits for lifetimes that support it. You may need to slightly tweak it depending on where you put it.
   ```csharp
       if (builder.Instance!.ApplicationLifetime is IControlledApplicationLifetime controlledApplicationLifetime) {
                    controlledApplicationLifetime.Exit += (sender, args) => {
                        ServiceManager.DisposeLocator();
                    };
                } 
   ```